kratos/utilities/sensitivity_builder.cpp

```diff
@@ -1,431 +1,317 @@
-    return false;
 }
 
-void AssembleOnDataValueContainer(const Variable<double>& rVariable,
-                                  const Vector& rSrc,
-                                  DataValueContainer& rDest)
+template <template <class T> class TFunctor, class TContainer, class... TArgs>
+void ExecuteFunctorInContainer(
+    const SensitivityBuilder::TSensitivityVariables& rVariables,
+    TContainer& rContainer,
+    TArgs&... rArgs)
 {
-    KRATOS_ERROR_IF(rSrc.size() != 1) << "Variable: " << rVariable.Name()
-                                      << ", rSrc.size() = " << rSrc.size() << std::endl;
-    rDest[rVariable] += rSrc[0];
+    block_for_each(rContainer, [&](typename TContainer::value_type& rEntity) {
+        ExecuteFunctor<TFunctor, typename TContainer::value_type, TArgs...>(
+            rVariables, rEntity, rArgs...);
+    });
 }
 
-void AssembleOnDataValueContainer(const Variable<array_1d<double, 3>>& rVariable,
-                                  const Vector& rSrc,
-                                  DataValueContainer& rDest)
+template <class TDataType>
+class SetNonHistoricalValueToZeroFunctor
 {
-    KRATOS_ERROR_IF(rSrc.size() > 3) << "Variable: " << rVariable.Name()
-                                     << ", rSrc.size() = " << rSrc.size() << std::endl;
-    array_1d<double, 3>& r_dest = rDest[rVariable];
-    for (std::size_t d = 0; d < rSrc.size(); ++d)
+public:
+    template <class TEntityType>
+    void operator()(
+        const SensitivityBuilder::SensitivityVariables<TDataType>& rVariable,
+        TEntityType& rEntity)
     {
-        r_dest[d] += rSrc[d];
+        rEntity.SetValue(*rVariable.pOutputVariable, rVariable.pOutputVariable->Zero());
     }
-}
+};
 
-void AssembleNodalSolutionStepValues(const Variable<double>& rVariable,
-                                     const Vector& rValues,
-                                     Geometry<Node<3>>& rGeom)
+template <class TDataType>
+class SetHistoricalValueToZeroFunctor
 {
-    KRATOS_TRY;
-    KRATOS_ERROR_IF(rGeom.size() != rValues.size())
-        << "Geometry size: " << rGeom.size()
-        << " is incompatible with vector size: " << rValues.size() << std::endl;
-    std::size_t index = 0;
-    for (auto& r_node : rGeom)
+public:
+    void operator()(
+        const SensitivityBuilder::SensitivityVariables<TDataType>& rVariable,
+        ModelPart::NodeType& rNode)
     {
-        if (r_node.GetValue(UPDATE_SENSITIVITIES))
-        {
-            double& r_dest = r_node.FastGetSolutionStepValue(rVariable);
-            const double rhs = rValues[index];
-#pragma omp atomic
-            r_dest += rhs;
-        }
-        ++index;
+        rNode.FastGetSolutionStepValue(*rVariable.pOutputVariable) =
+            rVariable.pOutputVariable->Zero();
     }
-    KRATOS_CATCH("");
+};
+
+template<class TDerivativeEntityType>
+bool HasSensitivityContributions(
+    const Geometry<ModelPart::NodeType>& rGeometry)
+{
+    return rGeometry.GetValue(UPDATE_SENSITIVITIES);
```

Comment:
Has... does a Get...?

Reply:
Since element and condition data value containers moved to geometry we have a Get :)

Reply:
ah ok now I understood what you do there